lib/Sema/DerivedConformanceDifferentiable.cpp

```diff
@@ -0,0 +1,524 @@
+      nominal, AutoDiffAssociatedVectorSpaceKind::Tangent);
+  switch (tangentStatus) {
+  case DoesNotExist:
+    TC.diagnose(derived.ConformanceDecl->getLoc(),
+                diag::differentiable_no_vector_space_struct,
+                derived.getProtocolType(), C.Id_TangentVector);
+    return nullptr;
+  case Invalid:
+    TC.diagnose(tangentDecl, diag::differentiable_invalid_vector_space_struct,
+                derived.getProtocolType(), C.Id_TangentVector);
+    return nullptr;
+  case Valid:
+    break;
+  }
+  auto tangentType = tangentDecl->getDeclaredInterfaceType();
+
+  auto *param =
+      new (C) ParamDecl(VarDecl::Specifier::Default, SourceLoc(), SourceLoc(),
+                        C.getIdentifier("along"), SourceLoc(),
+                        C.getIdentifier("direction"), parentDC);
+  param->setInterfaceType(tangentType);
+  ParameterList *params = ParameterList::create(C, {param});
+
+  DeclName declName(C, C.Id_moved, params);
+  auto funcDecl =
+      FuncDecl::create(C, SourceLoc(), StaticSpellingKind::None, SourceLoc(),
+                       declName, SourceLoc(),
+                       /*Throws*/ false, SourceLoc(),
+                       /*GenericParams=*/nullptr, params,
+                       TypeLoc::withoutLoc(selfInterfaceType), parentDC);
+  funcDecl->setImplicit();
+  funcDecl->setBodySynthesizer(deriveBodyDifferentiable_moved);
+
+  if (auto env = parentDC->getGenericEnvironmentOfContext())
+    funcDecl->setGenericEnvironment(env);
+  funcDecl->computeType();
+  funcDecl->copyFormalAccessFrom(nominal, /*sourceIsParentContext*/ true);
+  funcDecl->setValidationToChecked();
+
+  derived.addMembersToConformanceContext({funcDecl});
+  C.addSynthesizedDecl(funcDecl);
+
+  return funcDecl;
+}
+
+// Synthesize the `tangentVector(from:)` function declaration.
+static ValueDecl *
+deriveDifferentiable_tangentVector(DerivedConformance &derived) {
+  auto nominal = derived.Nominal;
+  auto &TC = derived.TC;
+  auto &C = derived.TC.Context;
+  auto parentDC = derived.getConformanceContext();
+
+  StructDecl *tangentDecl;
+  VectorSpaceStructStatus tangentStatus;
+  std::tie(tangentDecl, tangentStatus) = getVectorSpaceStructDecl(
+      nominal, AutoDiffAssociatedVectorSpaceKind::Tangent);
+  switch (tangentStatus) {
+  case DoesNotExist:
+    TC.diagnose(derived.ConformanceDecl->getLoc(),
+                diag::differentiable_no_vector_space_struct,
+                derived.getProtocolType(), C.Id_TangentVector);
+    return nullptr;
+  case Invalid:
+    TC.diagnose(tangentDecl, diag::differentiable_invalid_vector_space_struct,
+                derived.getProtocolType(), C.Id_TangentVector);
+    return nullptr;
+  case Valid:
+    break;
+  }
+  auto tangentType = tangentDecl->getDeclaredInterfaceType();
+
+  StructDecl *cotangentDecl;
+  VectorSpaceStructStatus cotangentStatus;
+  std::tie(cotangentDecl, cotangentStatus) = getVectorSpaceStructDecl(
```

Comment:
It'd be good to create a helper to wrap invocations of `getVectorSpaceStructDecl`. There are three invocations total, with the same switch block.